api/fees/data.go

```diff
@@ -83,14 +114,98 @@ func (fd *FeesData) getOrLoadFees(blockID thor.Bytes32) (*FeeCacheEntry, error)
 		return nil, err
 	}
 
-	header := block.Header()
+	var rewards *rewards
+	// If rewardPercentiles is not empty, we need to calculate the rewards for the block
+	if len(rewardPercentiles) > 0 {
+		rewards, err = fd.getRewardsForCache(block)
+		if err != nil {
+			return nil, err
+		}
+	}
 
+	header := block.Header()
 	fees = &FeeCacheEntry{
 		baseFee:       getBaseFee(header.BaseFee()),
 		gasUsedRatio:  float64(header.GasUsed()) / float64(header.GasLimit()),
+		cachedRewards: rewards,
 		parentBlockID: header.ParentID(),
 	}
 	fd.cache.Set(header.ID(), fees, float64(header.Number()))
 
 	return fees.(*FeeCacheEntry), nil
 }
+
+func (fd *FeesData) getRewardsForCache(block *block.Block) (*rewards, error) {
+	header := block.Header()
+	receipts, err := fd.repo.GetBlockReceipts(header.ID())
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the baseGasPrice for legacy transactions
+	parentSummary, err := fd.repo.GetBlockSummary(block.Header().ParentID())
+	if err != nil {
+		return nil, err
+	}
+	state := fd.stater.NewState(parentSummary.Root())
+
+	baseGasPrice, err := builtin.Params.Native(state).Get(thor.KeyBaseGasPrice)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the effective priority fee (reward) for each transaction
+	transactions := block.Transactions()
+	items := make([]rewardItem, len(transactions))
+	isGalactica := header.Number() >= thor.GetForkConfig(fd.repo.NewBestChain().GenesisID()).GALACTICA
+	for i, tx := range transactions {
+		provedWork, err := tx.ProvedWork(header.Number(), fd.repo.NewBestChain().GetBlockID)
+		if err != nil {
+			return nil, err
+		}
+		items[i] = rewardItem{
+			reward:  fork.GalacticaPriorityPrice(tx, baseGasPrice, provedWork, &fork.GalacticaItems{IsActive: isGalactica, BaseFee: header.BaseFee()}),
+			gasUsed: receipts[i].GasUsed,
+		}
+	}
+
+	// Sort by reward
+	slices.SortStableFunc(items, func(a, b rewardItem) int {
+		return a.reward.Cmp(b.reward)
+	})
+
+	return &rewards{
+		items:        items,
+		totalGasUsed: header.GasUsed(),
+	}, nil
+}
+
+func (fd *FeesData) calculateRewards(cachedRewards *rewards, rewardPercentiles []float64) []*hexutil.Big {
```

Comment:
Do you think it's worth to add some more unit tests for this function alone? Maybe with some edge cases like cachedRewards is nil, rewardPercentiles is [] and other examples

Reply:
Sure thing @paologalligit , there are currently integration tests that should cover indirectly both scenarios (essentially all tests at `fees_test.go` except for `TestRewardPercentiles` do not pass `rewardPercentiles`) but there is no harm in having unit tests also here, will do it.

Thanks.

Reply:
Done now